File: implementations/rust/src/years/y2022/day_5/part_2.rs
use crate::years::y2022::day_5::common::{parse_input, take_tops};

fn execute_commands(columns: &mut Vec<Vec<char>>, instructions: Vec<[usize; 3]>) {
    instructions.into_iter().for_each(|[n, from, to]| {
        let remaining_items = columns[from].len() - n;
        let mut top = columns[from].drain(remaining_items..).collect::<Vec<char>>();
        columns[to].append(&mut top);
    })
}

pub fn execute(input: String) -> String {
    let (mut columns, instructions) = parse_input(&input);

    execute_commands(&mut columns, instructions);

    return take_tops(&columns).iter().collect();
}

File: implementations/node/core/years/2022/day-5/part-2.mjs
import { executeInstructions, parseInput } from "./common.mjs";

export function execute(input) {
    const [columns, instructions] = parseInput(input)
    return executeInstructions(columns, instructions, (columns, [n, from, to]) => {
        let movedCrates = columns[from].splice(columns[from].length - n, n);
        columns[to].push(...movedCrates)
    })
}
